0.24.1: pytest fails in 13 unists

I'm packaging your module as an rpm package so I'm using the typical PEP517 based build, install and test cycle used on building packages from non-root account.
- `python3 -sBm build -w --no-isolation`
- because I'm calling `build` with `--no-isolation` I'm using during all processes only locally installed modules
- install .whl file in </install/prefix> using `installer` module
- run pytest with $PYTHONPATH pointing to sitearch and sitelib inside </install/prefix>
- build is performed in env which is *`cut off from access to the public network`* (pytest is executed with `-m "not network"`)

<details>
<summary>Here is pytest output:</summary>

```console
+ PYTHONPATH=/home/tkloczko/rpmbuild/BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib64/python3.10/site-packages:/home/tkloczko/rpmbuild/BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages
+ /usr/bin/pytest -ra -m 'not network'
============================= test session starts ==============================
platform linux -- Python 3.10.14, pytest-8.1.1, pluggy-1.4.0
rootdir: /home/tkloczko/rpmbuild/BUILD/pyudev-0.24.1
configfile: pytest.ini
plugins: hypothesis-6.100.0
collected 161 items

tests/test_core.py ........                                              [  4%]
tests/test_device.py F....F................F.......................s.... [ 36%]
...FF.F...s...                                                           [ 45%]
tests/test_discover.py ....                                              [ 47%]
tests/test_enumerate.py F.Fs.F.....                                      [ 54%]
tests/test_monitor.py ..Exception ignored in: <function Monitor.__del__ at 0x7feb11adcd30>
Traceback (most recent call last):
  File "/home/tkloczko/rpmbuild/BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages/pyudev/monitor.py", line 85, in __del__
    self._libudev.udev_monitor_unref(self)
ctypes.ArgumentError: argument 1: TypeError: expected LP_udev_monitor instance instead of _SentinelObject
....................FF                           [ 69%]
tests/test_observer.py .ssssss..ss                                       [ 76%]
tests/test_observer_deprecated.py ssssssssss.....sssss                   [ 88%]
tests/test_pypi.py s                                                     [ 89%]
tests/test_util.py ............FF...                                     [100%]

=================================== FAILURES ===================================
_________________________ TestAttributes.test_getitem __________________________

self = <tests._device_tests._attributes_tests.TestAttributes object at 0x7feb1078b490>

    @given(_CONTEXT_STRATEGY, strategies.sampled_from(_DEVICE_DATA))
>   @settings(max_examples=5)

tests/_device_tests/_attributes_tests.py:44:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <tests._device_tests._attributes_tests.TestAttributes object at 0x7feb1078b490>
a_context = <pyudev.core.Context object at 0x7feb11260520>
device_datum = DeviceData(/devices/pci0000:85/0000:85:1e.6)

    @given(_CONTEXT_STRATEGY, strategies.sampled_from(_DEVICE_DATA))
    @settings(max_examples=5)
    def test_getitem(self, a_context, device_datum):
        """
        Test that attribute value exists and is instance of bytes.
        """
        device = Devices.from_path(a_context, device_datum.device_path)
>       assert all(
            isinstance(device.attributes.get(key), bytes)
            for key in device_datum.attributes.keys()
        )
E       AssertionError
E       Falsifying example: test_getitem(
E           self=<tests._device_tests._attributes_tests.TestAttributes object at 0x7feb1078b490>,
E           a_context=<pyudev.core.Context object at 0x7feb11260520>,
E           device_datum=DeviceData(/devices/pci0000:85/0000:85:1e.6),
E       )
E       Explanation:
E           These lines were always and only run by failing examples:
E               /home/tkloczko/rpmbuild/BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages/pyudev/device/_device.py:1183

tests/_device_tests/_attributes_tests.py:50: AssertionError
__________________________ TestAttributes.test_asbool __________________________

self = <tests._device_tests._attributes_tests.TestAttributes object at 0x7feb1078b970>

    @given(_CONTEXT_STRATEGY, strategies.sampled_from(_DEVICE_DATA))
>   @settings(max_examples=5)

tests/_device_tests/_attributes_tests.py:109:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <tests._device_tests._attributes_tests.TestAttributes object at 0x7feb1078b970>
a_context = <pyudev.core.Context object at 0x7feb11260520>
device_datum = DeviceData(/devices/pci0000:85/0000:85:1e.6)

    @given(_CONTEXT_STRATEGY, strategies.sampled_from(_DEVICE_DATA))
    @settings(max_examples=5)
    def test_asbool(self, a_context, device_datum):
        """
        Test that bool result is a bool or ValueError raised.
        """
        device = Devices.from_path(a_context, device_datum.device_path)
        for key, value in device_datum.attributes.items():
            if value in ("0", "1"):
                try:
                    assert device.attributes.asbool(key) in (False, True)
                except KeyError:
                    pass
            else:
>               with pytest.raises(ValueError):
E               Failed: DID NOT RAISE <class 'ValueError'>
E               Falsifying example: test_asbool(
E                   self=<tests._device_tests._attributes_tests.TestAttributes object at 0x7feb1078b970>,
E                   a_context=<pyudev.core.Context object at 0x7feb11260520>,
E                   device_datum=DeviceData(/devices/pci0000:85/0000:85:1e.6),
E               )
E               Explanation:
E                   These lines were always and only run by failing examples:
E                       /home/tkloczko/rpmbuild/BUILD/pyudev-0.24.1/tests/_device_tests/_attributes_tests.py:126
E                       /home/tkloczko/rpmbuild/BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages/pyudev/device/_device.py:1183
E                       /usr/lib/python3.10/site-packages/_pytest/python_api.py:995

tests/_device_tests/_attributes_tests.py:122: Failed
________________________ TestDevice.test_device_number _________________________

self = <tests._device_tests._device_tests.TestDevice object at 0x7feb1078b5b0>

    @given(_CONTEXT_STRATEGY, strategies.sampled_from(_DEVICE_DATA))
>   @settings(max_examples=5)

tests/_device_tests/_device_tests.py:216:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/_device_tests/_device_tests.py:219: in test_device_number
    assert device.device_number == device_datum.device_number
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = DeviceData(/devices/platform/serial8250/serial8250:0/serial8250:0.30/tty/ttyS30)

    @property
    def device_number(self):
        """
        Get the device number as integer or 0 if the device has no device
        number.
        """
        device_node = self.device_node
>       return 0 if device_node is None else os.stat(device_node).st_rdev
E       FileNotFoundError: [Errno 2] No such file or directory: '/dev/ttyS30'
E       Falsifying example: test_device_number(
E           self=<tests._device_tests._device_tests.TestDevice object at 0x7feb1078b5b0>,
E           a_context=<pyudev.core.Context object at 0x7feb11260520>,
E           device_datum=DeviceData(/devices/platform/serial8250/serial8250:0/serial8250:0.30/tty/ttyS30),
E       )
E       Explanation:
E           These lines were always and only run by failing examples:
E               /home/tkloczko/rpmbuild/BUILD/pyudev-0.24.1/tests/utils/udev.py:109

tests/utils/udev.py:273: FileNotFoundError
----------------------------- Captured stderr call -----------------------------
No device node found: No such file or directory
No device node found: No such file or directory
No device node found: No such file or directory
No device node found: No such file or directory
No device node found: No such file or directory
_____________________ TestDevices.test_from_device_number ______________________

self = <tests._device_tests._devices_tests.TestDevices object at 0x7feb1079f460>

    @failed_health_check_wrapper
>   @given(
        _CONTEXT_STRATEGY,
        device_strategy(filter_func=lambda x: x.device_node is not None),
    )

tests/_device_tests/_devices_tests.py:129:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <tests._device_tests._devices_tests.TestDevices object at 0x7feb1079f460>
a_context = <pyudev.core.Context object at 0x7feb11260520>
a_device = Device('/sys/devices/LNXSYSTM:00/LNXPWRBN:00/input/input0/event0')

    @failed_health_check_wrapper
    @given(
        _CONTEXT_STRATEGY,
        device_strategy(filter_func=lambda x: x.device_node is not None),
    )
    @settings(max_examples=5)
    def test_from_device_number(self, a_context, a_device):
        """
        Verify that from_device_number() yields the correct device.
        """
>       mode = os.stat(a_device.device_node).st_mode
E       FileNotFoundError: [Errno 2] No such file or directory: '/dev/input/event0'
E       Falsifying example: test_from_device_number(
E           self=<tests._device_tests._devices_tests.TestDevices object at 0x7feb1079f460>,
E           a_context=<pyudev.core.Context object at 0x7feb11260520>,
E           a_device=Device('/sys/devices/LNXSYSTM:00/LNXPWRBN:00/input/input0/event0'),
E       )

tests/_device_tests/_devices_tests.py:138: FileNotFoundError
________________ TestDevices.test_from_device_number_wrong_type ________________

self = <tests._device_tests._devices_tests.TestDevices object at 0x7feb1079f2e0>

    @failed_health_check_wrapper
>   @given(
        _CONTEXT_STRATEGY,
        device_strategy(filter_func=lambda x: x.device_node is not None),
    )

tests/_device_tests/_devices_tests.py:144:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <tests._device_tests._devices_tests.TestDevices object at 0x7feb1079f2e0>
a_context = <pyudev.core.Context object at 0x7feb11260520>
a_device = Device('/sys/devices/LNXSYSTM:00/LNXPWRBN:00/input/input0/event0')

    @failed_health_check_wrapper
    @given(
        _CONTEXT_STRATEGY,
        device_strategy(filter_func=lambda x: x.device_node is not None),
    )
    @settings(max_examples=5)
    def test_from_device_number_wrong_type(self, a_context, a_device):
        """
        Verify appropriate behavior on real device number but swapped
        subsystems.
        """
>       mode = os.stat(a_device.device_node).st_mode
E       FileNotFoundError: [Errno 2] No such file or directory: '/dev/input/event0'
E       Falsifying example: test_from_device_number_wrong_type(
E           self=<tests._device_tests._devices_tests.TestDevices object at 0x7feb1079f2e0>,
E           a_context=<pyudev.core.Context object at 0x7feb11260520>,
E           a_device=Device('/sys/devices/LNXSYSTM:00/LNXPWRBN:00/input/input0/event0'),
E       )

tests/_device_tests/_devices_tests.py:154: FileNotFoundError
______________________ TestDevices.test_from_device_file _______________________
  + Exception Group Traceback (most recent call last):
  |   File "/usr/lib/python3.10/site-packages/_pytest/runner.py", line 340, in from_call
  |     result: Optional[TResult] = func()
  |   File "/usr/lib/python3.10/site-packages/_pytest/runner.py", line 240, in <lambda>
  |     lambda: runtest_hook(item=item, **kwds), when=when, reraise=reraise
  |   File "/usr/lib/python3.10/site-packages/pluggy/_hooks.py", line 501, in __call__
  |     return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  |   File "/usr/lib/python3.10/site-packages/pluggy/_manager.py", line 119, in _hookexec
  |     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  |   File "/usr/lib/python3.10/site-packages/pluggy/_callers.py", line 181, in _multicall
  |     return outcome.get_result()
  |   File "/usr/lib/python3.10/site-packages/pluggy/_result.py", line 99, in get_result
  |     raise exc.with_traceback(exc.__traceback__)
  |   File "/usr/lib/python3.10/site-packages/pluggy/_callers.py", line 166, in _multicall
  |     teardown.throw(outcome._exception)
  |   File "/usr/lib/python3.10/site-packages/_pytest/threadexception.py", line 87, in pytest_runtest_call
  |     yield from thread_exception_runtest_hook()
  |   File "/usr/lib/python3.10/site-packages/_pytest/threadexception.py", line 63, in thread_exception_runtest_hook
  |     yield
  |   File "/usr/lib/python3.10/site-packages/pluggy/_callers.py", line 166, in _multicall
  |     teardown.throw(outcome._exception)
  |   File "/usr/lib/python3.10/site-packages/_pytest/unraisableexception.py", line 90, in pytest_runtest_call
  |     yield from unraisable_exception_runtest_hook()
  |   File "/usr/lib/python3.10/site-packages/_pytest/unraisableexception.py", line 65, in unraisable_exception_runtest_hook
  |     yield
  |   File "/usr/lib/python3.10/site-packages/pluggy/_callers.py", line 166, in _multicall
  |     teardown.throw(outcome._exception)
  |   File "/usr/lib/python3.10/site-packages/_pytest/logging.py", line 849, in pytest_runtest_call
  |     yield from self._runtest_for(item, "call")
  |   File "/usr/lib/python3.10/site-packages/_pytest/logging.py", line 832, in _runtest_for
  |     yield
  |   File "/usr/lib/python3.10/site-packages/pluggy/_callers.py", line 166, in _multicall
  |     teardown.throw(outcome._exception)
  |   File "/usr/lib/python3.10/site-packages/_pytest/capture.py", line 883, in pytest_runtest_call
  |     return (yield)
  |   File "/usr/lib/python3.10/site-packages/pluggy/_callers.py", line 166, in _multicall
  |     teardown.throw(outcome._exception)
  |   File "/usr/lib/python3.10/site-packages/_pytest/skipping.py", line 256, in pytest_runtest_call
  |     return (yield)
  |   File "/usr/lib/python3.10/site-packages/pluggy/_callers.py", line 102, in _multicall
  |     res = hook_impl.function(*args)
  |   File "/usr/lib/python3.10/site-packages/_pytest/runner.py", line 182, in pytest_runtest_call
  |     raise e
  |   File "/usr/lib/python3.10/site-packages/_pytest/runner.py", line 172, in pytest_runtest_call
  |     item.runtest()
  |   File "/usr/lib/python3.10/site-packages/_pytest/python.py", line 1777, in runtest
  |     self.ihook.pytest_pyfunc_call(pyfuncitem=self)
  |   File "/usr/lib/python3.10/site-packages/pluggy/_hooks.py", line 501, in __call__
  |     return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  |   File "/usr/lib/python3.10/site-packages/pluggy/_manager.py", line 119, in _hookexec
  |     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  |   File "/usr/lib/python3.10/site-packages/pluggy/_callers.py", line 138, in _multicall
  |     raise exception.with_traceback(exception.__traceback__)
  |   File "/usr/lib/python3.10/site-packages/pluggy/_callers.py", line 102, in _multicall
  |     res = hook_impl.function(*args)
  |   File "/usr/lib/python3.10/site-packages/_pytest/python.py", line 200, in pytest_pyfunc_call
  |     result = testfunction(**testargs)
  |   File "/home/tkloczko/rpmbuild/BUILD/pyudev-0.24.1/tests/utils/misc.py", line 53, in the_func
  |     func(*args)
  |   File "/home/tkloczko/rpmbuild/BUILD/pyudev-0.24.1/tests/_device_tests/_devices_tests.py", line 179, in test_from_device_file
  |     @given(
  |   File "/usr/lib/python3.10/site-packages/hypothesis/core.py", line 1638, in wrapped_test
  |     raise the_error_hypothesis_found
  | exceptiongroup.ExceptionGroup: Hypothesis found 2 distinct failures. (2 sub-exceptions)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "/home/tkloczko/rpmbuild/BUILD/pyudev-0.24.1/tests/_device_tests/_devices_tests.py", line 188, in test_from_device_file
    |     device = Devices.from_device_file(a_context, a_device.device_node)
    |   File "/home/tkloczko/rpmbuild/BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages/pyudev/device/_device.py", line 226, in from_device_file
    |     return cls.from_device_number(context, device_type, device_number)
    |   File "/home/tkloczko/rpmbuild/BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages/pyudev/device/_device.py", line 183, in from_device_number
    |     raise DeviceNotFoundByNumberError(typ, number)
    | pyudev._errors.DeviceNotFoundByNumberError: No char device with number 34816
    | Falsifying example: test_from_device_file(
    |     self=<tests._device_tests._devices_tests.TestDevices object at 0x7feb1079efe0>,
    |     a_context=<pyudev.core.Context object at 0x7feb11260520>,
    |     a_device=Device('/sys/devices/virtual/tty/console'),
    | )
    +---------------- 2 ----------------
    | Traceback (most recent call last):
    |   File "/home/tkloczko/rpmbuild/BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages/pyudev/device/_device.py", line 221, in from_device_file
    |     device_type = get_device_type(filename)
    |   File "/home/tkloczko/rpmbuild/BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages/pyudev/_util.py", line 122, in get_device_type
    |     mode = os.stat(filename).st_mode
    | FileNotFoundError: [Errno 2] No such file or directory: '/dev/input/event0'
    |
    | During handling of the above exception, another exception occurred:
    |
    | Traceback (most recent call last):
    |   File "/home/tkloczko/rpmbuild/BUILD/pyudev-0.24.1/tests/_device_tests/_devices_tests.py", line 188, in test_from_device_file
    |     device = Devices.from_device_file(a_context, a_device.device_node)
    |   File "/home/tkloczko/rpmbuild/BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages/pyudev/device/_device.py", line 224, in from_device_file
    |     raise DeviceNotFoundByFileError(err)
    | pyudev._errors.DeviceNotFoundByFileError: [Errno 2] No such file or directory: '/dev/input/event0'
    | Falsifying example: test_from_device_file(
    |     self=<tests._device_tests._devices_tests.TestDevices object at 0x7feb1079efe0>,
    |     a_context=<pyudev.core.Context object at 0x7feb11260520>,
    |     a_device=Device('/sys/devices/LNXSYSTM:00/LNXPWRBN:00/input/input0/event0'),  # or any other generated value
    | )
    +------------------------------------
_____________________ TestEnumerator.test_match_subsystem ______________________

self = <tests.test_enumerate.TestEnumerator object at 0x7feb104fa680>

    @failed_health_check_wrapper
>   @given(_CONTEXT_STRATEGY, _SUBSYSTEM_STRATEGY)

tests/test_enumerate.py:90:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (<tests.test_enumerate.TestEnumerator object at 0x7feb104fa680>, <pyudev.core.Context object at 0x7feb11260520>, 'acpi')
kwargs = {}, arg_drawtime = 0.0006094439886510372, start = 312762.813466577
result = None, finish = 312763.189037809, in_drawtime = 0.0
runtime = datetime.timedelta(microseconds=375571)
current_deadline = timedelta(milliseconds=200)

    @proxies(self.test)
    def test(*args, **kwargs):
        arg_drawtime = math.fsum(data.draw_times.values())
        start = time.perf_counter()
        try:
            result = self.test(*args, **kwargs)
        finally:
            finish = time.perf_counter()
            in_drawtime = math.fsum(data.draw_times.values()) - arg_drawtime
            runtime = datetime.timedelta(seconds=finish - start - in_drawtime)
            self._timing_features = {
                "execute:test": finish - start - in_drawtime,
                **data.draw_times,
                **data._stateful_run_times,
            }

        if (current_deadline := self.settings.deadline) is not None:
            if not is_final:
                current_deadline = (current_deadline // 4) * 5
            if runtime >= current_deadline:
>               raise DeadlineExceeded(runtime, self.settings.deadline)
E               hypothesis.errors.DeadlineExceeded: Test took 375.57ms, which exceeds the deadline of 200.00ms
E               Falsifying example: test_match_subsystem(
E                   self=<tests.test_enumerate.TestEnumerator object at 0x7feb104fa680>,
E                   context=<pyudev.core.Context object at 0x7feb11260520>,
E                   subsystem='acpi',  # or any other generated value
E               )

/usr/lib/python3.10/site-packages/hypothesis/core.py:846: DeadlineExceeded
_____________ TestEnumerator.test_match_subsystem_nomatch_complete _____________

self = <tests.test_enumerate.TestEnumerator object at 0x7feb104faaa0>

    @failed_health_check_wrapper
>   @given(_CONTEXT_STRATEGY, _SUBSYSTEM_STRATEGY)

tests/test_enumerate.py:115:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (<tests.test_enumerate.TestEnumerator object at 0x7feb104faaa0>, <pyudev.core.Context object at 0x7feb11260520>, 'acpi')
kwargs = {}, arg_drawtime = 0.0006182260112836957, start = 312828.666100432
result = None, finish = 312829.246451372, in_drawtime = 0.0
runtime = datetime.timedelta(microseconds=580351)
current_deadline = timedelta(milliseconds=200)

    @proxies(self.test)
    def test(*args, **kwargs):
        arg_drawtime = math.fsum(data.draw_times.values())
        start = time.perf_counter()
        try:
            result = self.test(*args, **kwargs)
        finally:
            finish = time.perf_counter()
            in_drawtime = math.fsum(data.draw_times.values()) - arg_drawtime
            runtime = datetime.timedelta(seconds=finish - start - in_drawtime)
            self._timing_features = {
                "execute:test": finish - start - in_drawtime,
                **data.draw_times,
                **data._stateful_run_times,
            }

        if (current_deadline := self.settings.deadline) is not None:
            if not is_final:
                current_deadline = (current_deadline // 4) * 5
            if runtime >= current_deadline:
>               raise DeadlineExceeded(runtime, self.settings.deadline)
E               hypothesis.errors.DeadlineExceeded: Test took 580.35ms, which exceeds the deadline of 200.00ms
E               Falsifying example: test_match_subsystem_nomatch_complete(
E                   self=<tests.test_enumerate.TestEnumerator object at 0x7feb104faaa0>,
E                   context=<pyudev.core.Context object at 0x7feb11260520>,
E                   subsystem='acpi',  # or any other generated value
E               )

/usr/lib/python3.10/site-packages/hypothesis/core.py:846: DeadlineExceeded
__________________ TestEnumeratorMatchCombinations.test_match __________________

self = <tests.test_enumerate.TestEnumeratorMatchCombinations object at 0x7feb104f8640>

    @given(
>       _CONTEXT_STRATEGY,
        _SUBSYSTEM_STRATEGY,
        _SYSNAME_STRATEGY,
        _MATCH_PROPERTY_STRATEGY,
    )

tests/test_enumerate.py:173:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (<tests.test_enumerate.TestEnumeratorMatchCombinations object at 0x7feb104f8640>, <pyudev.core.Context object at 0x7feb11260520>, 'acpi', 'LNXSYSTM:00', ('DEVPATH', '/devices/LNXSYSTM:00'))
kwargs = {}, arg_drawtime = 0.0014324799412861466, start = 312970.172044585
result = None, finish = 312970.470424754, in_drawtime = 0.0
runtime = datetime.timedelta(microseconds=298380)
current_deadline = timedelta(milliseconds=200)

    @proxies(self.test)
    def test(*args, **kwargs):
        arg_drawtime = math.fsum(data.draw_times.values())
        start = time.perf_counter()
        try:
            result = self.test(*args, **kwargs)
        finally:
            finish = time.perf_counter()
            in_drawtime = math.fsum(data.draw_times.values()) - arg_drawtime
            runtime = datetime.timedelta(seconds=finish - start - in_drawtime)
            self._timing_features = {
                "execute:test": finish - start - in_drawtime,
                **data.draw_times,
                **data._stateful_run_times,
            }

        if (current_deadline := self.settings.deadline) is not None:
            if not is_final:
                current_deadline = (current_deadline // 4) * 5
            if runtime >= current_deadline:
>               raise DeadlineExceeded(runtime, self.settings.deadline)
E               hypothesis.errors.DeadlineExceeded: Test took 298.38ms, which exceeds the deadline of 200.00ms
E               Falsifying example: test_match(
E                   # The test always failed when commented parts were varied together.
E                   self=<tests.test_enumerate.TestEnumeratorMatchCombinations object at 0x7feb104f8640>,
E                   context=<pyudev.core.Context object at 0x7feb11260520>,
E                   subsystem='acpi',  # or any other generated value
E                   sysname='LNXSYSTM:00',  # or any other generated value
E                   ppair=('DEVPATH', '/devices/LNXSYSTM:00'),  # or any other generated value
E               )

/usr/lib/python3.10/site-packages/hypothesis/core.py:846: DeadlineExceeded
_________________ TestMonitorObserver.test_deprecated_handler __________________

self = <tests.test_monitor.TestMonitorObserver object at 0x7feb10515ae0>
fake_monitor = <tests.plugins.fake_monitor.FakeMonitor object at 0x7feb0fcd51b0>
fake_monitor_device = Device('/sys/devices/pci0000:00/0000:00:14.0/usb1/1-7/1-7.1/1-7.1:1.1')

    def test_deprecated_handler(self, fake_monitor, fake_monitor_device):
        observer = self.make_observer(fake_monitor, use_deprecated=True)
        observer.start()
        fake_monitor.trigger_event()
        fake_monitor.trigger_event()
        # wait a second for the tests to finish, and kill the observer if
        # it is still alive then
        observer.join(1)
        if observer.is_alive():
            observer.stop()
        assert not observer.is_alive()
>       assert self.events == [(None, fake_monitor_device)] * 2
E       AttributeError: 'TestMonitorObserver' object has no attribute 'events'

tests/test_monitor.py:297: AttributeError
________________________ TestMonitorObserver.test_fake _________________________

self = <tests.test_monitor.TestMonitorObserver object at 0x7feb105157b0>
fake_monitor = <tests.plugins.fake_monitor.FakeMonitor object at 0x7feb0fcc84c0>
fake_monitor_device = Device('/sys/devices/pci0000:00/0000:00:17.0/ata12/host11/scsi_host/host11')

    def test_fake(self, fake_monitor, fake_monitor_device):
        observer = self.make_observer(fake_monitor)
        observer.start()
        fake_monitor.trigger_event()
        fake_monitor.trigger_event()
        # wait a second for the tests to finish
        observer.join(1)
        # forcibly quit the thread if it is still alive
        if observer.is_alive():
            observer.stop()
        assert not observer.is_alive()
        # check that we got two events
>       assert self.events == [fake_monitor_device] * 2
E       AttributeError: 'TestMonitorObserver' object has no attribute 'events'

tests/test_monitor.py:311: AttributeError
____________________ test_get_device_type_character_device _____________________

    @pytest.mark.skipif(len(_CHAR_DEVICES) == 0, reason="no tty devices")
>   @given(strategies.sampled_from(_CHAR_DEVICES))

tests/test_util.py:123:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/test_util.py:129: in test_get_device_type_character_device
    assert _util.get_device_type(a_device.device_node) == "char"
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

filename = '/dev/ttyS0'

    def get_device_type(filename):
        """
        Get the device type of a device file.

        ``filename`` is a string containing the path of a device file.

        Return ``'char'`` if ``filename`` is a character device, or ``'block'`` if
        ``filename`` is a block device.  Raise :exc:`~exceptions.ValueError` if
        ``filename`` is no device file at all.  Raise
        :exc:`~exceptions.EnvironmentError` if ``filename`` does not exist or if
        its metadata was inaccessible.

        .. versionadded:: 0.15
        """
>       mode = os.stat(filename).st_mode
E       FileNotFoundError: [Errno 2] No such file or directory: '/dev/ttyS0'
E       Falsifying example: test_get_device_type_character_device(
E           a_device=Device('/sys/devices/platform/serial8250/serial8250:0/serial8250:0.0/tty/ttyS0'),
E       )

../../BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages/pyudev/_util.py:122: FileNotFoundError
______________________ test_get_device_type_block_device _______________________

    @pytest.mark.skipif(len(_BLOCK_DEVICES) == 0, reason="no block devices")
>   @given(strategies.sampled_from(_BLOCK_DEVICES))

tests/test_util.py:136:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/test_util.py:142: in test_get_device_type_block_device
    assert _util.get_device_type(a_device.device_node) == "block"
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

filename = '/dev/sdc'

    def get_device_type(filename):
        """
        Get the device type of a device file.

        ``filename`` is a string containing the path of a device file.

        Return ``'char'`` if ``filename`` is a character device, or ``'block'`` if
        ``filename`` is a block device.  Raise :exc:`~exceptions.ValueError` if
        ``filename`` is no device file at all.  Raise
        :exc:`~exceptions.EnvironmentError` if ``filename`` does not exist or if
        its metadata was inaccessible.

        .. versionadded:: 0.15
        """
>       mode = os.stat(filename).st_mode
E       FileNotFoundError: [Errno 2] No such file or directory: '/dev/sdc'
E       Falsifying example: test_get_device_type_block_device(
E           a_device=Device('/sys/devices/pci0000:00/0000:00:17.0/ata10/host9/target9:0:0/9:0:0:0/block/sdc'),
E       )

../../BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages/pyudev/_util.py:122: FileNotFoundError
=============================== warnings summary ===============================
tests/test_device.py::TestDevice::test_find_parent_no_devtype_mock
tests/test_device.py::TestDevice::test_find_parent_with_devtype_mock
  /usr/lib/python3.10/site-packages/_pytest/unraisableexception.py:80: PytestUnraisableExceptionWarning: Exception ignored in: <function Device.__del__ at 0x7feb11cc5510>

  Traceback (most recent call last):
    File "/home/tkloczko/rpmbuild/BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages/pyudev/device/_device.py", line 476, in __del__
      self._libudev.udev_device_unref(self)
  ctypes.ArgumentError: argument 1: TypeError: expected LP_udev_device instance instead of _SentinelObject

    warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))

tests/test_monitor.py::TestMonitor::test_from_netlink_source_udev_mock
tests/test_monitor.py::TestMonitor::test_from_netlink_source_kernel_mock
  /usr/lib/python3.10/site-packages/_pytest/unraisableexception.py:80: PytestUnraisableExceptionWarning: Exception ignored in: <function Monitor.__del__ at 0x7feb11adcd30>

  Traceback (most recent call last):
    File "/home/tkloczko/rpmbuild/BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages/pyudev/monitor.py", line 85, in __del__
      self._libudev.udev_monitor_unref(self)
  ctypes.ArgumentError: argument 1: TypeError: expected LP_udev_monitor instance instead of _SentinelObject

    warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))

tests/test_monitor.py::TestMonitorObserver::test_deprecated_handler
  /usr/lib/python3.10/site-packages/_pytest/threadexception.py:77: PytestUnhandledThreadExceptionWarning: Exception in thread Thread-1

  Traceback (most recent call last):
    File "/usr/lib64/python3.10/threading.py", line 1016, in _bootstrap_inner
      self.run()
    File "/home/tkloczko/rpmbuild/BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages/pyudev/monitor.py", line 544, in run
      self._callback(device)
    File "/home/tkloczko/rpmbuild/BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages/pyudev/monitor.py", line 518, in <lambda>
      callback = lambda d: event_handler(d.action, d)
    File "/home/tkloczko/rpmbuild/BUILD/pyudev-0.24.1/tests/test_monitor.py", line 262, in event_handler
      self.events.append((action, device))
  AttributeError: 'TestMonitorObserver' object has no attribute 'events'

    warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))

tests/test_monitor.py::TestMonitorObserver::test_fake
  /usr/lib/python3.10/site-packages/_pytest/threadexception.py:77: PytestUnhandledThreadExceptionWarning: Exception in thread Thread-2

  Traceback (most recent call last):
    File "/usr/lib64/python3.10/threading.py", line 1016, in _bootstrap_inner
      self.run()
    File "/home/tkloczko/rpmbuild/BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages/pyudev/monitor.py", line 544, in run
      self._callback(device)
    File "/home/tkloczko/rpmbuild/BUILD/pyudev-0.24.1/tests/test_monitor.py", line 257, in callback
      self.events.append(device)
  AttributeError: 'TestMonitorObserver' object has no attribute 'events'

    warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))

tests/test_observer.py::TestGlibObserver::test_monitor
  /home/tkloczko/rpmbuild/BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages/pyudev/glib.py:121: PyGIDeprecationWarning: GObject.SIGNAL_RUN_LAST is deprecated; use GObject.SignalFlags.RUN_LAST instead
    GObject.SIGNAL_RUN_LAST,

tests/test_observer.py::TestGlibObserver::test_monitor
  /home/tkloczko/rpmbuild/BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages/pyudev/glib.py:154: PyGIDeprecationWarning: GObject.SIGNAL_RUN_LAST is deprecated; use GObject.SignalFlags.RUN_LAST instead
    GObject.SIGNAL_RUN_LAST,

tests/test_observer.py::TestGlibObserver::test_monitor
  /home/tkloczko/rpmbuild/BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages/pyudev/glib.py:159: PyGIDeprecationWarning: GObject.SIGNAL_RUN_LAST is deprecated; use GObject.SignalFlags.RUN_LAST instead
    GObject.SIGNAL_RUN_LAST,

tests/test_observer.py::TestGlibObserver::test_monitor
  /home/tkloczko/rpmbuild/BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages/pyudev/glib.py:164: PyGIDeprecationWarning: GObject.SIGNAL_RUN_LAST is deprecated; use GObject.SignalFlags.RUN_LAST instead
    GObject.SIGNAL_RUN_LAST,

tests/test_observer.py::TestGlibObserver::test_monitor
  /home/tkloczko/rpmbuild/BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages/pyudev/glib.py:169: PyGIDeprecationWarning: GObject.SIGNAL_RUN_LAST is deprecated; use GObject.SignalFlags.RUN_LAST instead
    GObject.SIGNAL_RUN_LAST,

tests/test_observer.py::TestGlibObserver::test_monitor
  /home/tkloczko/rpmbuild/BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages/pyudev/glib.py:174: PyGIDeprecationWarning: GObject.SIGNAL_RUN_LAST is deprecated; use GObject.SignalFlags.RUN_LAST instead
    GObject.SIGNAL_RUN_LAST,

tests/test_observer.py::TestGlibObserver::test_events_fake_monitor
  /home/tkloczko/rpmbuild/BUILD/pyudev-0.24.1/tests/test_observer.py:159: Warning: Source ID 5 was not found when attempting to remove it
    self.glib.source_remove(source)

tests/test_observer_deprecated.py::TestDeprecatedGlibObserver::test_monitor
tests/test_observer_deprecated.py::TestDeprecatedGlibObserver::test_events_fake_monitor[add]
tests/test_observer_deprecated.py::TestDeprecatedGlibObserver::test_events_fake_monitor[remove]
tests/test_observer_deprecated.py::TestDeprecatedGlibObserver::test_events_fake_monitor[change]
tests/test_observer_deprecated.py::TestDeprecatedGlibObserver::test_events_fake_monitor[move]
  /home/tkloczko/rpmbuild/BUILDROOT/python-pyudev-0.24.1-8.fc37.x86_64/usr/lib/python3.10/site-packages/pyudev/glib.py:186: DeprecationWarning: Will be removed in 1.0. Use pyudev.glib.MonitorObserver instead.
    warnings.warn(

tests/test_observer_deprecated.py::TestDeprecatedGlibObserver::test_events_fake_monitor[add]
  /home/tkloczko/rpmbuild/BUILD/pyudev-0.24.1/tests/test_observer_deprecated.py:174: Warning: Source ID 11 was not found when attempting to remove it
    self.glib.source_remove(source)

tests/test_observer_deprecated.py::TestDeprecatedGlibObserver::test_events_fake_monitor[remove]
  /home/tkloczko/rpmbuild/BUILD/pyudev-0.24.1/tests/test_observer_deprecated.py:174: Warning: Source ID 15 was not found when attempting to remove it
    self.glib.source_remove(source)

tests/test_observer_deprecated.py::TestDeprecatedGlibObserver::test_events_fake_monitor[change]
  /home/tkloczko/rpmbuild/BUILD/pyudev-0.24.1/tests/test_observer_deprecated.py:174: Warning: Source ID 19 was not found when attempting to remove it
    self.glib.source_remove(source)

tests/test_observer_deprecated.py::TestDeprecatedGlibObserver::test_events_fake_monitor[move]
  /home/tkloczko/rpmbuild/BUILD/pyudev-0.24.1/tests/test_observer_deprecated.py:174: Warning: Source ID 23 was not found when attempting to remove it
    self.glib.source_remove(source)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
SKIPPED [1] tests/_device_tests/_device_tests.py:474: unsafe to check ID_WWN_WITH_EXTENSION
SKIPPED [1] tests/_device_tests/_tags_tests.py:50: no device with tags
SKIPPED [1] tests/utils/misc.py:55: failed health check for test_match_property_bool() (/home/tkloczko/rpmbuild/BUILD/pyudev-0.24.1/tests/test_enumerate.py: 132)
SKIPPED [1] tests/test_observer.py:99: could not import 'PySide.QtCore': No module named 'PySide'
SKIPPED [1] tests/test_observer.py:104: could not import 'PySide.QtCore': No module named 'PySide'
SKIPPED [1] tests/test_observer.py:99: could not import 'PyQt4.QtCore': No module named 'PyQt4'
SKIPPED [1] tests/test_observer.py:104: could not import 'PyQt4.QtCore': No module named 'PyQt4'
SKIPPED [1] tests/test_observer.py:99: could not import 'PyQt5.QtCore': No module named 'PyQt5'
SKIPPED [1] tests/test_observer.py:104: could not import 'PyQt5.QtCore': No module named 'PyQt5'
SKIPPED [1] tests/test_observer.py:99: Display required for wxPython
SKIPPED [1] tests/test_observer.py:104: Display required for wxPython
SKIPPED [1] tests/test_observer_deprecated.py:90: could not import 'PySide.QtCore': No module named 'PySide'
SKIPPED [4] tests/test_observer_deprecated.py:95: could not import 'PySide.QtCore': No module named 'PySide'
SKIPPED [1] tests/test_observer_deprecated.py:90: could not import 'PyQt4.QtCore': No module named 'PyQt4'
SKIPPED [4] tests/test_observer_deprecated.py:95: could not import 'PyQt4.QtCore': No module named 'PyQt4'
SKIPPED [1] tests/test_observer_deprecated.py:90: Display required for wxPython
SKIPPED [4] tests/test_observer_deprecated.py:95: Display required for wxPython
SKIPPED [1] tests/test_pypi.py:38: Not in git clone
FAILED tests/test_device.py::TestAttributes::test_getitem - AssertionError
FAILED tests/test_device.py::TestAttributes::test_asbool - Failed: DID NOT RA...
FAILED tests/test_device.py::TestDevice::test_device_number - FileNotFoundErr...
FAILED tests/test_device.py::TestDevices::test_from_device_number - FileNotFo...
FAILED tests/test_device.py::TestDevices::test_from_device_number_wrong_type
FAILED tests/test_device.py::TestDevices::test_from_device_file - exceptiongr...
FAILED tests/test_enumerate.py::TestEnumerator::test_match_subsystem - hypoth...
FAILED tests/test_enumerate.py::TestEnumerator::test_match_subsystem_nomatch_complete
FAILED tests/test_enumerate.py::TestEnumeratorMatchCombinations::test_match
FAILED tests/test_monitor.py::TestMonitorObserver::test_deprecated_handler - ...
FAILED tests/test_monitor.py::TestMonitorObserver::test_fake - AttributeError...
FAILED tests/test_util.py::test_get_device_type_character_device - FileNotFou...
FAILED tests/test_util.py::test_get_device_type_block_device - FileNotFoundEr...
===== 13 failed, 121 passed, 27 skipped, 22 warnings in 314.26s (0:05:14) ======
```
</details>

<details>
<summary>List of installed modules in build env:</summary>

```console
Package                       Version
----------------------------- -----------
alabaster                     0.7.16
attrs                         23.2.0
Babel                         2.14.0
build                         1.2.1
charset-normalizer            3.3.2
docutils                      0.20.1
exceptiongroup                1.1.3
hypothesis                    6.100.0
idna                          3.7
imagesize                     1.4.1
importlib_metadata            7.1.0
iniconfig                     2.0.0
installer                     0.7.0
Jinja2                        3.1.3
MarkupSafe                    2.1.5
packaging                     24.0
pluggy                        1.4.0
Pygments                      2.17.2
PyGObject                     3.48.2
pyproject_hooks               1.0.0
pytest                        8.1.1
python-dateutil               2.9.0.post0
requests                      2.31.0
setuptools                    69.4.0
snowballstemmer               2.2.0
sortedcontainers              2.4.0
Sphinx                        7.2.6
sphinxcontrib-applehelp       1.0.8
sphinxcontrib-devhelp         1.0.6
sphinxcontrib-htmlhelp        2.0.5
sphinxcontrib-jsmath          1.0.1
sphinxcontrib-qthelp          1.0.7
sphinxcontrib-serializinghtml 1.1.10
tokenize_rt                   5.2.0
tomli                         2.0.1
urllib3                       2.2.1
wheel                         0.43.0
zipp                          3.18.1
```
</details>

Please let me know if you need more details or want me to perform some diagnostics.

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -19,15 +19,15 @@ jobs:
         include:
           # MANDATORY CHECKS USING CURRENT DEVELOPMENT INTERPRETER
           - python-version: "3.12"
-            dependencies: pytest=="7" hypothesis
+            dependencies: pytest hypothesis
             task: PYTHONPATH=./src make -f Makefile test-travis
           # MANDATORY CHECKS USING LOWEST SUPPORTED INTERPRETER
           - python-version: "3.10"
-            dependencies: pytest=="7" hypothesis
+            dependencies: pytest hypothesis
             task: PYTHONPATH=./src make -f Makefile test-travis
           # MANDATORY CHECKS USING PYPY INTERPRETER
           - python-version: pypy-3.9
-            dependencies: pytest=="7" hypothesis
+            dependencies: pytest hypothesis
             task: PYTHONPATH=./src make -f Makefile test-travis
     runs-on: ubuntu-latest
     steps:

--- a/tests/test_monitor.py
+++ b/tests/test_monitor.py
@@ -277,10 +277,10 @@ class TestMonitorObserver:
             self.observer = MonitorObserver(monitor, callback=self.callback)
         return self.observer
 
-    def setup(self):
+    def setup_method(self):
         self.events = []
 
-    def teardown(self):
+    def teardown_method(self):
         self.events = None
 
     def test_deprecated_handler(self, fake_monitor, fake_monitor_device):

